msg-transport/src/tcp_tls/mod.rs

```diff
@@ -0,0 +1,275 @@
+use core::fmt;
+use derive_more::{Deref, DerefMut, From};
+use futures::future::BoxFuture;
+use openssl::ssl::{Ssl, SslAcceptor, SslConnector, SslMethod};
+use std::{
+    io,
+    net::SocketAddr,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tokio::net::{TcpListener, TcpStream};
+use tokio_openssl::SslStream;
+use tracing::debug;
+
+use msg_common::async_error;
+
+use crate::{Acceptor, PeerAddress, Transport, TransportExt};
+
+pub mod config;
+
+/// An invalid operation, due to using as a server or viceversa.
+#[derive(Debug, Clone, Copy, thiserror::Error)]
+pub enum InvalidOperation {
+    #[error("tried to bind as a client")]
+    BindAsClient,
+    #[error("tried to connect as a server")]
+    ConnectAsServer,
+    #[error("tried to accept a connection as a client")]
+    AcceptAsClient,
+}
+
+/// Error that can occur using the TCP transport with TLS.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("i/o error: {0}")]
+    Io(#[from] io::Error),
+    #[error("i/o error kind: {0}")]
+    IoKind(io::ErrorKind),
+    /// OpenSSL library error.
+    #[error("openssl error: {0}")]
+    OpenSsl(#[from] openssl::error::ErrorStack),
+    /// SSL connection error.
+    #[error("ssl error: {0}")]
+    Ssl(#[from] openssl::ssl::Error),
+    #[error("invalid configuration: {0}")]
+    InvalidOperation(#[from] InvalidOperation),
+}
+
+/// A TCP-TLS client.
+#[derive(Debug, Clone, From, Deref, DerefMut)]
+pub struct Client(config::Client);
+
+/// A TCP-TLS server.
+pub struct Server {
+    /// The underlying TCP listener.
+    listener: Option<TcpListener>,
+    /// The OpenSSL acceptor for TLS handshake requests.
+    acceptor: Arc<SslAcceptor>,
+}
+
+impl fmt::Debug for Server {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Server")
+            .field("listener", &self.listener)
+            .field("acceptor", &"SslAcceptor")
+            .finish()
+    }
+}
+
+/// A TCP transport implementation with TLS support using OpenSSL.
+#[derive(Debug, From)]
+pub enum TcpTls {
+    Client(Client),
+    Server(Server),
+}
+
+impl TcpTls {
+    /// Create a new instance of the transport with client configuration.
+    pub fn new_client(config: config::Client) -> Self {
+        TcpTls::Client(config.into())
```

Comment:
Similar to the above, different pattern between client (into) and server (inline)